execstats: refactor TraceAnalyzer stats calculations

cathymw opened the pull request:
Based on this suggestion: https://github.com/cockroachdb/cockroach/pull/57495#pullrequestreview-546519364
This commit refactors TraceAnalyzer to have a single function that processes
stats and stores them in the TraceAnalyzer, reducing the duplicated code
used for calculating stats.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57711)
<!-- Reviewable:end -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-:-MO48bDR5tJo2MaZ3NZE:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @cathymw, and @RaduBerinde)

___
*[pkg/sql/execstats/traceanalyzer.go, line 172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MO485p74KUieUsYx2IU:-MO485p74KUieUsYx2IV:b-ueb453) ([raw file](https://github.com/cockroachdb/cockroach/blob/d22259256143f5ca3d17ce3f7b457f4130e7f5c3/pkg/sql/execstats/traceanalyzer.go#L172)):*
> ```Go
> // while calculating stats, ProcessStats adds the error to a slice to be returned
> // to the caller, but continues calculating other stats.
> func (a *TraceAnalyzer) ProcessStats() []error {
> ```

Consider returning a single `error` instead of `[]error`. You can use errors.CombineErrors or errors.WithSecondaryError to combine multiple errors into a single one.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57711)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @cathymw, and @RaduBerinde)

___
*[pkg/sql/execstats/traceanalyzer.go, line 94 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOBGF-BANeeAD827udU:-MOBGF-BANeeAD827udV:buq5r4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/d22259256143f5ca3d17ce3f7b457f4130e7f5c3/pkg/sql/execstats/traceanalyzer.go#L94)):*
> ```Go
> 	*FlowMetadata
> 	NetworkBytesSentGroupedByNode map[roachpb.NodeID]int64
> 	MaxMemoryUsageFromFlowStats   int64
> ```

I don't think it makes sense to keep this and the field below separate. Additionally, I think we've heard from customers that it's nice to be able to break down these stats by node, so I would make this a `map[rochpb.NodeID]int64` grouped by node as above.

What might make sense is putting these fields into a new `NodeLevelStats` struct similar to `QueryLevelStats` and returning that in a `GetNodeLevelStats` method. The nice thing is that you avoid the possibility of modification by callers (probably good practice) by returning a value.

Maybe `ProcessStats` should populate two fields, the node and query level stats at the same time, and these stats would be accessed through `GetQueryLevelStats` and `GetNodeLevelStats`. What do you think?
___
*[pkg/sql/execstats/traceanalyzer.go, line 172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MO485p74KUieUsYx2IU:-MOBHA8h9vwaCTKK4HKN:bbadrd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d22259256143f5ca3d17ce3f7b457f4130e7f5c3/pkg/sql/execstats/traceanalyzer.go#L172)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Consider returning a single `error` instead of `[]error`. You can use errors.CombineErrors or errors.WithSecondaryError to combine multiple errors into a single one.
</blockquote></details>

+1, didn't think about this.


<!-- Sent from Reviewable.io -->

cathymw committed:
execstats: refactor TraceAnalyzer stats calculations

This commit refactors TraceAnalyzer to have a single function that processes
stats and stores them in the TraceAnalyzer, reducing the duplicated code
used for calculating stats.

Release note: None.

cathymw reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57711)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/execstats/traceanalyzer.go, line 94 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOBGF-BANeeAD827udU:-MOEjQn915qDD6gXPYVE:b-4amo3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/d22259256143f5ca3d17ce3f7b457f4130e7f5c3/pkg/sql/execstats/traceanalyzer.go#L94)):*
>I don't think it makes sense to keep this and the field below separate. Additionally, I think we've heard from customers that it's nice to be able to break down these stats by node, so I would make this a `map[rochpb.NodeID]int64` grouped by node as above.

Done.

>What might make sense is putting these fields into a new `NodeLevelStats` struct similar to `QueryLevelStats` and returning that in a `GetNodeLevelStats` method. The nice thing is that you avoid the possibility of modification by callers (probably good practice) by returning a value.

Done.

>Maybe `ProcessStats` should populate two fields, the node and query level stats at the same time, and these stats would be accessed through `GetQueryLevelStats` and `GetNodeLevelStats`. What do you think?

I like the idea of `ProcessStats` populating the node level stats and being able to access these stats through `GetNodeLevelStats`. But I personally think that having the `ProcessStats` function also populate the `QueryLevelStats` would be making the function do too many things, since it would have to use the node level stats it just calculated to calculate the query level stats. 
I ended up renaming the processing function to `ProcessNodeLevelStats` so that it's more clear about which stats it's calculating
___
*[pkg/sql/execstats/traceanalyzer.go, line 172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MO485p74KUieUsYx2IU:-MOEjO0e62cd0cfv9-zV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d22259256143f5ca3d17ce3f7b457f4130e7f5c3/pkg/sql/execstats/traceanalyzer.go#L172)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

+1, didn't think about this.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

cathymw reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57711)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/execstats/traceanalyzer.go, line 94 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOBGF-BANeeAD827udU:-MOIQDHa9RGQocRzMP3p:b66a81y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d22259256143f5ca3d17ce3f7b457f4130e7f5c3/pkg/sql/execstats/traceanalyzer.go#L94)):*
<details><summary><i>Previously, cathymw wrote…</i></summary><blockquote>

>I don't think it makes sense to keep this and the field below separate. Additionally, I think we've heard from customers that it's nice to be able to break down these stats by node, so I would make this a `map[rochpb.NodeID]int64` grouped by node as above.

Done.

>What might make sense is putting these fields into a new `NodeLevelStats` struct similar to `QueryLevelStats` and returning that in a `GetNodeLevelStats` method. The nice thing is that you avoid the possibility of modification by callers (probably good practice) by returning a value.

Done.

>Maybe `ProcessStats` should populate two fields, the node and query level stats at the same time, and these stats would be accessed through `GetQueryLevelStats` and `GetNodeLevelStats`. What do you think?

I like the idea of `ProcessStats` populating the node level stats and being able to access these stats through `GetNodeLevelStats`. But I personally think that having the `ProcessStats` function also populate the `QueryLevelStats` would be making the function do too many things, since it would have to use the node level stats it just calculated to calculate the query level stats. 
I ended up renaming the processing function to `ProcessNodeLevelStats` so that it's more clear about which stats it's calculating
</blockquote></details>

Nevermind, I thought about it some more and I think what you initially suggested makes sense. `ProcessStats` now calls `processNodeLevelStats` and `processQueryLevelStats` to calculate the stats and populate the two fields. 


<!-- Sent from Reviewable.io -->

asubiotto reviewed (changes requested):
> Reviewed 2 of 11 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57711)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @cathymw)

___
*[pkg/sql/instrumentation.go, line 192 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOW4TYl2Too94VHNtfd:-MOW4TYl2Too94VHNtfe:b74v0yi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac76e488606e6e755fedf79519406e2de2a956ac/pkg/sql/instrumentation.go#L192)):*
> ```Go
> 		}
> 		queryLevelStats, errors := execstats.GetQueryLevelStats(trace, cfg.TestingKnobs.DeterministicExplainAnalyze, flowMetadata)
> 		if errors != nil {
> ```

nit: `s/errors/error`
___
*[pkg/sql/execstats/traceanalyzer.go, line 179 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOW8TTI2LiH5svSeB07:-MOW8TTI2LiH5svSeB08:b-o4mcia) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac76e488606e6e755fedf79519406e2de2a956ac/pkg/sql/execstats/traceanalyzer.go#L179)):*
> ```Go
> func (a *TraceAnalyzer) ProcessNodeLevelStats() error {
> 	networkBytesSentGroupedByNode := make(map[roachpb.NodeID]int64)
> 	maxMemoryUsageGroupedByNode := make(map[roachpb.NodeID]int64)
> ```

nit: I would set the a.nodeLevelStats maps here directly and update them directly below. It avoids new additions having to create a new map and then remember to set it to the field later on.
___
*[pkg/sql/execstats/traceanalyzer.go, line 215 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOW9TiJ0-krgSRyO8UF:-MOW9TiJ0-krgSRyO8UG:b7s35yc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac76e488606e6e755fedf79519406e2de2a956ac/pkg/sql/execstats/traceanalyzer.go#L215)):*
> ```Go
> 		// The vectorized flow attaches the MaxMemUsage stat to a flow level span, so we need to check
> 		// flow stats for max memory usage.
> 		// TODO(cathymw): maxMemUsage shouldn't be attached to span stats that are associated with streams,
> ```

nit: should this TODO live with the code above that calculates mem usage from stream stats?
___
*[pkg/sql/execstats/traceanalyzer.go, line 268 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOW4mbhEUtjROmWJA0J:-MOW4mbhEUtjROmWJA0K:b-8ofj1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac76e488606e6e755fedf79519406e2de2a956ac/pkg/sql/execstats/traceanalyzer.go#L268)):*
> ```Go
> // GetQueryLevelStats tries to process as many stats as possible. If errors occur
> // while processing stats, GetQueryLevelStats returns the combined errors to the caller
> //// but continues calculating other stats.
> ```

nit: `////`
___
*[pkg/sql/execstats/traceanalyzer_test.go, line 170 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOW4vh3BvVyQ01OZA4P:-MOW4vh3BvVyQ01OZA4Q:b-9ckml3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac76e488606e6e755fedf79519406e2de2a956ac/pkg/sql/execstats/traceanalyzer_test.go#L170)):*
> ```Go
> 			},
> 		} {
> 			nodeLevelStats := tc.analyzer.GetNodeLevelStats()
> ```

I would call `GetQueryLevelStats` here and above and check the value returned by that. It also verifies that the logic in `GetQueryLevelStats` is correct.


<!-- Sent from Reviewable.io -->

cathymw reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57711)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/instrumentation.go, line 192 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOW4TYl2Too94VHNtfd:-MOXeL8Y2R_GzxZ-0nbX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac76e488606e6e755fedf79519406e2de2a956ac/pkg/sql/instrumentation.go#L192)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: `s/errors/error`
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer.go, line 179 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOW8TTI2LiH5svSeB07:-MOXexdR9aebniT8e5Xt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac76e488606e6e755fedf79519406e2de2a956ac/pkg/sql/execstats/traceanalyzer.go#L179)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: I would set the a.nodeLevelStats maps here directly and update them directly below. It avoids new additions having to create a new map and then remember to set it to the field later on.
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer.go, line 215 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOW9TiJ0-krgSRyO8UF:-MOXf7lW1zN7wh7Ytmyo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac76e488606e6e755fedf79519406e2de2a956ac/pkg/sql/execstats/traceanalyzer.go#L215)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: should this TODO live with the code above that calculates mem usage from stream stats?
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer.go, line 268 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOW4mbhEUtjROmWJA0J:-MOXfBVK-Rm80LfW6YPE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac76e488606e6e755fedf79519406e2de2a956ac/pkg/sql/execstats/traceanalyzer.go#L268)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: `////`
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer_test.go, line 170 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOW4vh3BvVyQ01OZA4P:-MOXiUGiBIp4Faf_bt2b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac76e488606e6e755fedf79519406e2de2a956ac/pkg/sql/execstats/traceanalyzer_test.go#L170)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I would call `GetQueryLevelStats` here and above and check the value returned by that. It also verifies that the logic in `GetQueryLevelStats` is correct.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (changes requested):
> Reviewed 2 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57711)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @cathymw)

___
*[pkg/sql/execstats/traceanalyzer.go, line 263 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOa6IPE-4Z98Nok0ggJ:-MOa6IPE-4Z98Nok0ggK:b7k7tu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc45309a812ac7987099a20a38a54aa969200906/pkg/sql/execstats/traceanalyzer.go#L263)):*
> ```Go
> func (a *TraceAnalyzer) ProcessStats() error {
> 	err := a.processNodeLevelStats()
> 	a.processQueryLevelStats()
> ```

It's invalid to call this before `processNodeLevelStats`, right? I would just move the `processNodeLevelSTats` and `processQueryLevelStats` logic into this method so you always process both.
___
*[pkg/sql/execstats/traceanalyzer_test.go, line 125 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOa6eEf9omxvKQQWhn4:-MOa6eEf9omxvKQQWhn5:b-kgfbqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc45309a812ac7987099a20a38a54aa969200906/pkg/sql/execstats/traceanalyzer_test.go#L125)):*
> ```Go
> 		analyzer := <-analyzerChan
> 		require.NoError(t, analyzer.AddTrace(trace, true /* makeDeterministic */))
> 		require.Empty(t, analyzer.ProcessStats())
> ```

nit: `s/Empty/NoError`


<!-- Sent from Reviewable.io -->

cathymw reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57711)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/execstats/traceanalyzer.go, line 263 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOa6IPE-4Z98Nok0ggJ:-MOb41eFCNYxaH_BLVbd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc45309a812ac7987099a20a38a54aa969200906/pkg/sql/execstats/traceanalyzer.go#L263)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It's invalid to call this before `processNodeLevelStats`, right? I would just move the `processNodeLevelSTats` and `processQueryLevelStats` logic into this method so you always process both.
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer_test.go, line 125 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-MOa6eEf9omxvKQQWhn4:-MOb42YIAQrOdWisZqyl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc45309a812ac7987099a20a38a54aa969200906/pkg/sql/execstats/traceanalyzer_test.go#L125)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: `s/Empty/NoError`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57711#-:-MOf0J_1EyvQ6HvVLQoz:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)



<!-- Sent from Reviewable.io -->

cathymw commented:
TFTRs!
bors r=asubiotto
